### ISSUE TITLE
lib/logstorage:   reduce the time consumption of `tokenizeStrings` function

### ISSUE THREAD

**Issue by z-anshun** [VictoriaMetrics/VictoriaMetrics#5041]
### Is your feature request related to a problem? Please describe

Hello, 
When I was testing VictoriaLogs, I found that the `tokenizeStrings` function took up a lot of resources, as shown in the flame graph below. Then I found that the [tokenizer.reset ](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/master/lib/logstorage/tokenizer.go#L132-L139)method would traverse the entire `map`, and because it was a hotspot function,it took up a lot of resources. Too much increases the load on the CPU.
Even if I update golang to version 1.21 and uuse the `clear` method to replace the `delete` method, there will not be a big improvement.

flame graph:
<img width="1353" alt="image" src="https://github.com/VictoriaMetrics/VictoriaMetrics/assets/57032282/4f721dcc-0ef6-4552-932e-5c2dbe87a839">

### Describe the solution you'd like

Maybe we don't have to actively clean up `map` resources and hand it over to golang's GC, which will achieve better results. Like this:
```go
func (t *tokenizer) reset() {
	t.m = map[string]struct{}{}
}
```
Updated flame graph:
<img width="1333" alt="image" src="https://github.com/VictoriaMetrics/VictoriaMetrics/assets/57032282/4bc17b73-1610-478d-b3dc-d7911efc4f6a">

**Comment by zekker6** [comment]
Hello @z-anshun, thank you for the detailed report!

Switching `tokenizer.reset` to allocate new map every time might increase memory usage & increase time spent by GC.
Any chance you've tested this in a long run and can share some results?

**Comment by z-anshun** [comment]
@zekker6 Okay, I found during testing that changing [tokenizer.reset](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/master/lib/logstorage/tokenizer.go#L38-L43) to allocate a new `map` each time can improve the peak performance of `VictoriaLogs` writes. However, it can become more unstable.

before update:
<img width="1367" alt="image" src="https://github.com/VictoriaMetrics/VictoriaMetrics/assets/57032282/cfabde54-48c8-467a-b425-9b28027e0916">

updated:
<img width="1379" alt="image" src="https://github.com/VictoriaMetrics/VictoriaMetrics/assets/57032282/e9924153-89b8-41d2-b869-5def0ba66f3b">

**Comment by zekker6** [comment]
@z-anshun It is very likely that Golang runtime starts doing GC which leads to drops in ingestion performance after some time. If you'd like to confirm this you can take CPU profiles and check if GC is taking a lot of CPU time during ingestion drops.

**Comment by zekker6** [comment]
@z-anshun Thanks for the details, that makes sense. Also, since you're running version which is built manually I can suggest you to create a build from the latest master branch as there were several updates related to background merges processing. (see https://github.com/VictoriaMetrics/VictoriaMetrics/commit/7b33a27874285c494156e7005095960c9801800b, https://github.com/VictoriaMetrics/VictoriaMetrics/commit/da9ef90277aa173deb5c5b29e463b29b4d66ddec)

**Comment by z-anshun** [comment]
@zekker6 Hi, I used the latest version of VictoriaLogs and found that the [tokenizer.reset](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/master/lib/logstorage/tokenizer.go#L38) method can still improve its performance by allocating a new `map`. Maybe the `map` has too many keys and the load on the cpu is too high.

before:
<img width="1363" alt="image" src="https://github.com/VictoriaMetrics/VictoriaMetrics/assets/57032282/3e07d125-873a-44ae-989e-11ae9b3ae971">
<img width="1362" alt="image" src="https://github.com/VictoriaMetrics/VictoriaMetrics/assets/57032282/cbeddb84-f54a-4032-9717-5c604e45b432">

updated:
<img width="1367" alt="image" src="https://github.com/VictoriaMetrics/VictoriaMetrics/assets/57032282/9b032cf7-7b3b-40af-b489-3fb756f5dced">
<img width="1364" alt="image" src="https://github.com/VictoriaMetrics/VictoriaMetrics/assets/57032282/7e2d21e8-26dd-4656-a13b-896b56397fcd">

**Comment by f41gh7** [comment]
I think, it's possible to implement own version of Set for token tracking.

**Comment by f41gh7** [comment]
According to the given profiles and graphs, a lot of time is wasted at `memclr` functions, which are CPU specific. I assume, that vmlogs launched at modern amd64 CPU, is it correct?

**Comment by z-anshun** [comment]
> According to the given profiles and graphs, a lot of time is wasted at `memclr` functions, which are CPU specific. I assume, that vmlogs launched at modern amd64 CPU, is it correct?

@f41gh7 Yes, here is some cpu information.

```
$ lscpu
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              96
On-line CPU(s) list: 0-95
Thread(s) per core:  2
Core(s) per socket:  24
Socket(s):           2
NUMA node(s):        2
Vendor ID:           GenuineIntel
CPU family:          6
Model:               85
Model name:          Intel(R) Xeon(R) Gold 6271C CPU @ 2.60GHz
Stepping:            7
CPU MHz:             3099.999
CPU max MHz:         2601.0000
CPU min MHz:         1000.0000
BogoMIPS:            5200.00
Virtualization:      VT-x
L1d cache:           32K
L1i cache:           32K
L2 cache:            1024K
L3 cache:            33792K
```

**Comment by f41gh7** [comment]
It seems to be an issue with map overflow buckets. In case of hash collisions, overflow bucket is used. Overflow bucket never deleted and it increases CPU time for it's clear.

  one possible option - recreate map after sometime ( few minutes) or after some number of re-usages.

**Comment by z-anshun** [comment]
> It seems to be an issue with map overflow buckets. In case of hash collisions, overflow bucket is used. Overflow bucket never deleted and it increases CPU time for it's clear.
> 
> one possible option - recreate map after sometime ( few minutes) or after some number of re-usages.

right! Too many elements in the map cause the bucket to overflow. Perhaps this map should not be reused directly. When it has too many elements, a new one can be created directly, as shown below. 
```go
func (t *tokenizer) reset() {
	if len(t.m) < 1024 {
		m := t.m
		for k := range m {
			delete(m, k)
		}
		return
	}
	t.m = map[string]struct{}{}
}
```
But I don’t know if this threshold is correct. Do you have any better suggestions or methods?
